cli/mender-artifact/debugfs.go

```diff
@@ -104,53 +105,100 @@ func debugfsReplaceFile(imageFile, newFile, image string) (err error) {
+	var deleteCmd string
+	if isDir {
+		return debugfsRemoveDir(imageFile, image, recursive)
+	} else {
+		deleteCmd = "rm"
+	}
+	cmd = fmt.Sprintf("%s %s\nclose", deleteCmd, imageFile)
+	_, err = executeCommand(cmd, image)
+	return err
+}
+
+func debugfsRemoveDir(imageFile, image string, recursive bool) (err error) {
+	// Remove the `/` suffix if present, as debugfs does not play nice with it.
+	imageFile = strings.TrimRight(imageFile, "/")
+	cmd := fmt.Sprintf("rmdir %s", imageFile)
+	if _, err = executeCommand(cmd, image); err != nil {
+		if recursive && strings.Contains(err.Error(), "directory not empty") {
+			// Recurse and remove all files in the directory.
+			cmd = fmt.Sprintf("ls %s", imageFile)
+			buf, err := executeCommand(cmd, image)
+			if err != nil {
+				return errors.Wrap(err, "debugfsRemoveDir")
+			}
+			regexp := regexp.MustCompile(`[0-9]+ +\([0-9]+\) +((.\w+.)+)`)
+			for _, filename := range regexp.FindAllStringSubmatch(buf.String(), -1) {
+				err = debugfsRemoveFileOrDir(filepath.Join(imageFile, string(filename[1])), image, recursive)
+				if err != nil {
+					return errors.Wrap(err, "debugfsRemoveDir")
+				}
+			}
+			// Now that all the subdirs and files should be removed, try and remove the directory
+			// once more.
+			return debugfsRemoveDir(imageFile, image, recursive)
+		}
+		return errors.Wrap(err, "debugfsRemoveDir")
+	}
+	return nil
 }
 
 // executeCommand takes a command string and passes it on to debugfs on the image given.
-func executeCommand(cmdstr, image string) error {
+func executeCommand(cmdstr, image string) (stdout *bytes.Buffer, err error) {
 	scr, err := ioutil.TempFile("", "mender-debugfs")
 	if err != nil {
-		return errors.Wrap(err, "debugfs: create sync script file")
+		return nil, errors.Wrap(err, "debugfs: create sync script file")
 	}
 	defer os.Remove(scr.Name())
 	defer scr.Close()
 
 	err = scr.Chmod(0755)
 	if err != nil {
-		return errors.Wrap(err, "debugfs: set script file exec flag")
+		return nil, errors.Wrap(err, "debugfs: set script file exec flag")
 	}
 
 	if _, err = scr.WriteString(cmdstr); err != nil {
-		return errors.Wrap(err, "debugfs: store sync script file")
+		return nil, errors.Wrap(err, "debugfs: store sync script file")
 	}
 
 	if err = scr.Close(); err != nil {
-		return errors.Wrap(err, "debugfs: close sync script")
+		return nil, errors.Wrap(err, "debugfs: close sync script")
 	}
 	cmd := exec.Command("debugfs", "-w", "-f", scr.Name(), image)
-	ep, _ := cmd.StderrPipe()
+	cmd.Env = []string{"DEBUGFS_PAGER='cat'"}
+	errbuf := bytes.NewBuffer(nil)
+	stdout = bytes.NewBuffer(nil)
+	cmd.Stderr = errbuf
+	cmd.Stdout = stdout
 	if err = cmd.Start(); err != nil {
-		return errors.Wrap(err, "debugfs: run debugfs script")
+		return nil, errors.Wrap(err, "debugfs: run debugfs script")
 	}
-	data, err := ioutil.ReadAll(ep)
-	if err != nil {
-		return err
+	if err := cmd.Wait(); err != nil {
+		return nil, err
 	}
```

Comment:
I think it is better to use `cmd.Run()` in this case instead of Start+Wait here, once there is no handling from our side of stderr/stdout